### ISSUE TITLE
[srt-live-transnmit] Output jitter

### ISSUE THREAD

**Issue by afgbloch** [Haivision/srt#2302]
Hello,

I have been making tests with the srt-live-transmit app to transmit live MPEG TS and have been experiencing jitter on the UDP output.

The tests have been conducted on a local wired network using dektek StreamXpress for the transmission and StreamXpert for the analysis of the TS on the other side. The jitter does not seem to come from the UDP source as going through the local network without the SRT library in the middle doesn't cause any issues.

Using the command as follow:

`./srt-live-transmit "udp://:1234" "srt://192.168.13.10:4201?latency=80" -v -s 8000`
`./srt-live-transmit "srt://:4201?latency=80" "udp://192.168.13.40:1234" -v -s 8000`

I need to add a 13ms buffer at the input of the StreamXpert for the TS to be stable. After reading  #1337 I have tried adding the options below but I have the same problem.

`./srt-live-transmit -srctime:yes -buffering:1 "udp://:1234" "srt://192.168.13.10:4201?latency=80" -v -s 8000`
`./srt-live-transmit -srctime:yes -buffering:1 "srt://:4201?latency=80" "udp://192.168.13.40:1234" -v -s 8000`

While if I use the srt-test-live app I do not detect any jitter above 1ms.

`./srt-test-live "udp://:1234" "srt://192.168.13.10:4201?latency=80" `
`./srt-test-live "srt://:4201?latency=80" "udp://192.168.13.40:1234" `

Is this something expected? Are there other options I should be aware of to reduce jitter on srt-live-transmit ?
Thanks

**Comment by maxsharabayko** [comment]
Hi @afgbloch 
`srt-live-transmit` is a sample application intended to be simple by design to also serve as an example. It shows how to unify handling SRT and UDP using the epoll concept. However, the `epoll` on UDP seems to be not as efficient as a simple `::select(..)`. It is described in [this section](https://github.com/Haivision/srt/blob/v1.4.4/docs/apps/srt-live-transmit.md#udp-performance) of the srt-live-transmit app doc.

Both `srt-test-live` (enabled with `-DENABLE_TESTING=ON` CMake build option) and `srt-xtransmit` ([link](https://github.com/maxsharabayko/srt-xtransmit)) use `::select` for that purpose, and should be considered as alternative for `srt-live-transmit` when doing this kinds of testing.

**Comment by maxsharabayko** [comment]
Closing as a duplicate of #1337

**Comment by ethouris** [comment]
Just to be precise - `srt-test-live` doesn't use `::select`. It normally uses only the "blocking mode" and optionally it can use the non-blocking mode with SRT epoll, only for SRT and only if requested (`blocking=no` should be specified in the URI parameters), which is only for testing purposes.

**Comment by maxsharabayko** [comment]
> Just to be precise - `srt-test-live` doesn't use `::select`. It normally uses only the "blocking mode" and optionally it can use the non-blocking mode with SRT epoll, only for SRT and only if requested (`blocking=no` should be specified in the URI parameters), which is only for testing purposes.

Right, UDP reading in `srt-test-live` is always a blocking operation then. Still better performance compared to epoll waiting in `srt-live-transmit`.

**Comment by ethouris** [comment]
Actually the SRT epoll waiting is doing the following thing:
1. Wakes up once per 10us (or something like that).
2. It can additionally wake up earlier if there's a signal coming from SRT internals when ANY socket has changed its state (regardless if the epoll ID used by the currently waiting function is subscribed for it or not).
3. When woken up, it checks readiness flags for subscribed SRT sockets, and - if there are any system sockets subscribed - calls an appropriate system socket waiting function (such as `::select`) with these system sockets subscribed and with zero timeout (that is, it should not wait for readiness). If nothing is ready, it repeats the steps, unless the timeout has elapsed, in which case it exits.

This means, more-less: For UDP it can fit in some range of bitrates. If the bitrate is too high, and you rely on the check before every call to `::read`, your app may read slower than the data are coming (busy waiting may help here a bit). If too low, then you  will waste CPU time on idle rolls of the waiting function.

**Comment by maxsharabayko** [comment]
Still, the main performance issue observed is in using `::epoll_wait(..)` for a UDP socket  in [CEpoll::wait(..)](https://github.com/Haivision/srt/blob/v1.4.4/srtcore/epoll.cpp#L642).
`srt-live-transmit` relies on SRT epoll and `srt_epoll_wait()` for polling UDP socket, and it is likely the cause of these jitter issues when the source stream comes via UDP.

Might actually be worth replacing `::epoll_wait(..)` with `::select(..)`  and checking if there is any performance improvement. 🤔

**Comment by maxsharabayko** [comment]
@afgbloch Could you re-check jitter using `srt-live-transmit` from [this branch](https://github.com/maxsharabayko/srt/tree/experiment/epoll-wait-udp-select) in my fork?
I left only the `::select` version of the SRT epoll waiting on a UDP socket. I wonder if you see any difference?

You platform is Linux, right?

**Comment by maxsharabayko** [comment]
Ah, sorry, I mixed up two different issues. The thing about `::select` is related to high bitrates.

This jitter problem is a bit different and is indeed deeper. As described [here](https://github.com/Haivision/srt/issues/2302#issuecomment-1104007533), SRT epoll does not actually wait for UDP and SRT sockets in parallel. It just checks UDP socket readiness with 0 ms timeout once it is woken up by a 10 ms timeout or a ready SRT socket. That is the source of the jitter on a UDP input of srt-live-transmit.


> I have also tried `srt-xtransmit` and the jitter issue is not there. Could this app be considered for usage in production compare to `srt-test-live`?

Both are dedicated to testing purposes. If any of those fits your needs, feel free to use themt. But at your own risk. The same is true for `srt-live-transmit` actually.

**Comment by maxsharabayko** [comment]
Closing this again as a duplicate of #1337. 🙂 
SRT epoll dedicated epic #2303 is created to track epoll improvements.
